Add Microsoft.AspNetCore.OpenApi to the webapiaot template (with an optional flag to disable it)

sander1095 opened the pull request:
# Add Microsoft.AspNetCore.OpenApi to the webapiaot template (with an optional flag to disable it)

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Creates feature parity with the webapi template in regards to OpenAPI

## Description

This PR adds `Microsoft.AspNetCore.OpenApi` to the `dotnet new webapiaot` template as .NET 9 made that package AOT compatible. Just like the `webapi` template, OpenAPI support is included by default and it now also contains a `--no-openapi` flag to disable OpenAPI if desired.

Fixes #59564

sander1095 committed:
Add basic set up for the new flag with translations by Copilot

sander1095 committed:
Add conditional logic to use OpenAPI

sander1095 committed:
Add OpenAPI test to WebAPIAOT template tests

sander1095 committed:
Fix tests

sander1095 committed:
Add missing endif in the csproj

sander1095 committed:
Add more metadata to get 404 not found in the openapi doc correctly

sander1095 committed:
Also add the 404 metadata to the other file

sander1095 commented on src/ProjectTemplates/Web.ProjectTemplates/content/WebApiAot-CSharp/Program.cs:
```diff
@@ -18,11 +30,24 @@
+    #if (EnableOpenAPI)
+    .WithName("GetTodoById")
+    .Produces<Todo>(StatusCodes.Status200OK)
+    .Produces(StatusCodes.Status404NotFound);
```
Discussion for the team:

Should we also add `ProblemDetails` to the list of `JsonSerializable` items at the bottom of the file?

At this time it's not needed, but if someone were to call `ProducesProblem` or `ProducesValidationProblem`, the user would get the following exception:

```
 NotSupportedException: JsonTypeInfo metadata for type 'Microsoft.AspNetCore.Mvc.ProblemDetails' was not provided by TypeInfoResolver of type 'System.Text.Json.Serialization.Metadata.JsonTypeInfoResolverWithAddedModifiers'. If using source generation, ensure that all root types passed to the serializer have been annotated with 'JsonSerializableAttribute', along with any types that might be serialized polymorphically. 
```

------------

I believe this is a bit intimidating. Adding ProblemDetails at the bottom (which most modern API's would use, I believe) would avoid these issues and provide a smoother experience.

Thoughts? :)

captainsafia replied:
> At this time it's not needed, but if someone were to call ProducesProblem or ProducesValidationProblem, the user would get the following exception:

Yes, this is an extremely annoying problem. We're tracking approaches to fix this via https://github.com/dotnet/aspnetcore/issues/56021.

For now, I think it's sufficient to keep the scope of the JsonSerializerContext in the templates to only the types that are directly used in the templates.

> I can create a seperate issue for this if the team is interested to explore this further.

I don't believe we have an issue for this on the backlog from taking a look so feel free to open one. I can post my thoughts on whether or not we should do this there.

sander1095 replied:
I've created https://github.com/dotnet/aspnetcore/issues/60394 to discuss this further. I'm resolving this comment chain for now!

captainsafia commented on src/ProjectTemplates/Web.ProjectTemplates/content/WebApiAot-CSharp/.template.config/localize/templatestrings.cs.json:
The localizations here are automated via our localization process so we shouldn't need to change anything except the *.en.json process.

sander1095 replied:
That saves me some time on Duolingo courses! 😉 I've reverted all the changes in the `templatestrings` files, except for the english one.

sander1095 replied:
Actually, when running `src\ProjectTemplates\build.cmd -test -NoRestore -NoBuild -NoBuildDeps -configuration Release` the files all get edited and get the english variant included again.

This leads me to believe this behaviour is intended, and I should include the english versions in all files which will then get picked up the localization process later.

Let me know if this is wrong; I can simply revert the changes again.

captainsafia commented on src/ProjectTemplates/Web.ProjectTemplates/content/WebApiAot-CSharp/Program.Main.cs:
```diff
@@ -24,11 +24,24 @@ public static void Main(string[] args)
+                : Results.NotFound())
+            #if (EnableOpenAPI)
+            .WithName("GetTodoById")
+            .Produces<Todo>(StatusCodes.Status200OK)
```
I think instead of using the `Produces` extension methods in the templates we should cosndier using the `TypedResults` extension methods and `IResult<>` return type to demonstrate these types of union-based returns.

sander1095 replied:
Done. We're now using TypedResults. I've chosen to keep using TypedResults even when the user disables OpenAPI:

- Using Results/TypedResults based on the fact if the user is using OpenAPI or not might cause confusion
- TypedResults are always better than Results, because of the added metadata and compile-time safety

Let me know if you agree, or if you want me to return to using `Results` in openapi-disabled scenarios.

captainsafia commented on src/ProjectTemplates/Web.ProjectTemplates/content/WebApiAot-CSharp/Program.Main.cs:
```diff
@@ -24,11 +24,24 @@ public static void Main(string[] args)
+            .Produces<Todo>(StatusCodes.Status200OK)
+            .Produces(StatusCodes.Status404NotFound);
+            #elif
+            ;
```
I think this semicolon need to be on L37.

sander1095 committed:
Change elif to else

sander1095 committed:
Remove unnecessary 118n changes

sander1095 committed:
Add english translation back

sander1095 committed:
Keep the semicolon on the same line as requested by PR review

sander1095 committed:
Use TypedResults instead of extension methods and reduce the if else statements

sander1095 committed:
Add template strings back after running `bash build.sh -test -configuration Release`.

I assume this is part of the natural process of creating a PR and thus should be committed.

captainsafia commented:
@sander1095 This looks good to me overall!

I'd like to get @DamianEdwards thoughts on this since he's got more experience with making template changes and might have thoughts on other things we need to do here. He's OOF right now so we might have to wait a while for his feedback.

Rick-Anderson commented:
@DamianEdwards  can you schedule time for review?

captainsafia commented:
/azp run

sander1095 commented:
Hi @DamianEdwards ! Thanks for approving this PR. Safia said:

> I'd like to get @DamianEdwards thoughts on this since he's got more experience with making template changes and might have thoughts on other things we need to do here. He's OOF right now so we might have to wait a while for his feedback.

Like I said [in my comments](https://github.com/dotnet/aspnetcore/pull/60337#issuecomment-2653676473) in this PR, I can not get the tests to run correctly, nor can I create a project to test my changes locally. 

Are you able to help out with this?

captainsafia commented:
@sander1095 Absent running tests locally, we can rely on test failiures in CI to diagnose issues. Based on the [build analysis](https://github.com/dotnet/aspnetcore/pull/60337/checks?check_run_id=38023410980), it seems like the templates are currently unbuildable because the correct usings for the TypedResult static methods are not in place.

sander1095 committed:
Merge branch 'main' into GH-59564-webapiaot-openapi

sander1095 committed:
Add usings for TypedResults and Results/NotFound/Ok etc..

sander1095 committed:
Add OpenAPI services to Program.Main.cs

sander1095 commented:
> @sander1095 Absent running tests locally, we can rely on test failiures in CI to diagnose issues. Based on the [build analysis](https://github.com/dotnet/aspnetcore/pull/60337/checks?check_run_id=38023410980), it seems like the templates are currently unbuildable because the correct usings for the TypedResult static methods are not in place.

That's odd! When creating a new project (`dotnet new webapiaot`) and adding `TypedResults/Results<T1,T2>`, no usings need to be added because `<ImplicitUsings>` is `True`, which is still the case in this PR.

I've added the 2 usings necessary for the project to compile when `ImplicitUsings` is `false`. Let's see if the pipelines progress further with these explicit changes.

-----

I also noticed I completely forgot to add `builder.Services.AddOpenApi();` and `app.MapOpenApi();` to `Program.Main.cs` 🤦‍♂️ . I've just added a commit for this, too. I'm curious for the pipeline results!

sander1095 committed:
Remove http using because thats already part of the implicit usings

captainsafia commented:
> What do you think, @captainsafia ? If you also agree with option 2, I can remove the draft status of the PR in the SDK repo

I think for this case, we can stick to explicit usings in the Program.cs for this templates. Modifying the global usings is a more global (heh) change to be mindful and it's harder to take back once we've done it compared to a local change in a single template.

sander1095 committed:
TEMP COMMIT - Removing EnableOpenAPI flags to find out why tests fail

sander1095 committed:
Revert "TEMP COMMIT - Removing EnableOpenAPI flags to find out why tests fail"

This reverts commit 63cf302e75f871ae01f0e77f84284a23ecdb28f4.

sander1095 committed:
Fix assertions where we expected OpenAPI to be enabled for published apps, which is not the case

sander1095 committed:
Added consistent indentation in test methods

sander1095 committed:
Fix build errors

captainsafia reviewed (approved):
LGTM! TY for adding tests and resolving all the details.

I kicked CI -- looked like a flake.
